daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala

```diff
@@ -0,0 +1,426 @@
+        val depth1 = depth.incr(1)
+        val env1 = trackBindings(depth, env, 1)
+        Bounce(() =>
+          transformExp(depth1, env1, body, { body1 =>
+            Bounce(() => txK(AExpr(SELet1(rhs, body1.wrapped))))
+          })(transform))
+    })
+  }
+
+  private[this] def flattenAlts[A](depth: DepthA, env: Env, alts: Array[SCaseAlt])(
+      k: K[Array[SCaseAlt], A]): Trampoline[A] = {
+    // Note: this could be made properly CPS and thus constant stack through
+    // trampoline by implementing a CPS version of map. However, map on an
+    // array is implemented as a loop so this should be fine.
+    Bounce(() =>
+      k(alts.map {
+        case SCaseAlt(pat, body0) =>
+          val n = patternNArgs(pat)
+          val env1 = trackBindings(depth, env, n)
+          flattenExp(depth.incr(n), env1, body0)(body => {
+            Land(SCaseAlt(pat, body.wrapped))
+          }).bounce
+      }))
+  }
+
+  private[this] def patternNArgs(pat: SCasePat): Int = pat match {
+    case _: SCPEnum | _: SCPPrimCon | SCPNil | SCPDefault | SCPNone => 0
+    case _: SCPVariant | SCPSome => 1
+    case SCPCons => 2
+  }
+
+  /** `transformExp` is the function at the heart of the ANF transformation.
+    *  You can read its type as saying: "Caller, give me a general expression
+    *  `exp`, (& depth/env info), and a transformation function `transform`
+    *  which says what you want to do with the transformed expression. Then I
+    *  will do the transform, and call `transform` with it. I reserve the right
+    *  to wrap further expression-AST around the expression returned by
+    *  `transform`.
+    *
+    *  See: `atomizeExp` for an instance where this wrapping occurs.
+    *
+    *  Note: this wrapping is the reason why we need a "second" CPS transform to
+    *  achieve constant stack through trampoline.
+    */
+  private[this] def transformExp[A](depth: DepthA, env: Env, exp: SExpr, k: K[AExpr, A])(
+      transform: Tx[SExpr, A]): Trampoline[A] =
+    exp match {
+      case atom0: SExprAtomic =>
+        val atom = makeRelativeA(depth)(makeAbsoluteA(env, atom0))
+        Bounce(() => transform(depth, atom, k))
+
+      case x: SEVal => Bounce(() => transform(depth, x, k))
+      case x: SEImportValue => Bounce(() => transform(depth, x, k))
+
+      case SEAppGeneral(func, args) =>
+        // It's safe to perform ANF if the func-expression has no effects when evaluated.
+        val safeFunc =
+          func match {
+            // we know that trivially in these two cases
+            case SEBuiltin(b) => (args.size <= b.arity)
+            case _: SEBuiltinRecursiveDefinition => (args.size <= 3)
+            case _ => false
+          }
+        // It's also safe to perform ANF for applications of a single argument.
+        if (safeFunc || args.size == 1) {
+          transformMultiApp[A](depth, env, func, args, k)(transform)
+        } else {
+          // Otherwise we expand the multi-app into a cascade of single-apps.
```

Comment:
I'm not convinced that we want to do this. Although it is correct it might be pretty bad for performance of executing the resulting AST since every application of a single argument needs to make a copy of the PAP and extend the array of arguments with the new arguments. The advantage of the old `KArg` was that we cloned the PAP once and the modified the array of arguments in place. We could do this because we knew we "owned" the array. The ability to gain this information of "ownership" is lost during this syntactic transformation if I'm not mistaken.

Reply:
This cascade of single apps only occurs in the 20% or so of cases when we can't be sure that the original ANF is safe. I'm not sure there is very much difference between the way a cascade of single-apps in executed (repeated PAP;apply) and the way a non-ANF multi-app is executed (repeated PUSH-continuation;Enter). But I am willing to believe the multi-app approach might be slightly more efficient. However the cost of the multi-app approach is the loss of linearity/simplicity in the ANF form, and the increased code-complexity of the engine

Reply:
The percentage of applications that fall into this category is not the only metric that matters here. It also matters how many arguments get applied. Unfortunately, applying `n` arguments now has a runtime complexity of `O(n^2)` since you need to copy the intermediate argument arrays for each argument and they grow with each argument added. I would prefer to not have any avoidable quadratic complexities in the heart of our engine. Thus, I'm _strongly_ in favour of adding a bit more complexity to the engine to avoid this.

Reply:
We agree to reinstate the non-ANF applications back into Speedy. Well, the func will still be atomic.